fix #1281 by adding support for DELETE endpoint on server

anurse opened the pull request:
Add support for a `DELETE` endpoint to stop the long polling connection gracefully (rather than just waiting for it to time out).

TODO:
* [x] C# Client support including Functional Testing
* [x] TS Client support
* [x] Functional testing of TS

JamesNK commented on src/Microsoft.AspNetCore.Http.Connections.Client/Internal/LongPollingTransport.cs:
```diff
@@ -185,6 +185,24 @@ public Task StartAsync(Uri url, TransferFormat transferFormat)
+            try
+            {
+                Log.SendingDeleteRequest(_logger, pollUrl);
+                var resp = await _httpClient.DeleteAsync(pollUrl);
```
response. Abbreviations trigger me.

anurse replied:
I'm on board with you. For some reason I keep doing it with `resp` though...

JamesNK commented on src/Microsoft.AspNetCore.Http.Connections.Client/Internal/LongPollingTransport.Log.cs:
```diff
@@ -99,6 +108,30 @@ public static void PollResponseReceived(ILogger logger, HttpResponseMessage resp
+            {
+                if (logger.IsEnabled(LogLevel.Debug))
+                {
+                    _sendingDeleteRequest(logger, pollUrl.ToString(), null);
```
Does the logging not ToString Uri's automatically? I'm sure I have seen logging that passes the Uri directly.

davidfowl replied:
Yep, we should be doing that.

davidfowl commented on src/Microsoft.AspNetCore.Http.Connections.Client/Internal/LongPollingTransport.cs:
```diff
@@ -185,6 +185,24 @@ public Task StartAsync(Uri url, TransferFormat transferFormat)
                 Log.ReceiveStopped(_logger);
+
+                // Send the DELETE request to clean-up the connection on the server, but don't wait for it
+                _ = SendDeleteRequest(pollUrl);
```
Why not await this?

davidfowl commented on test/Microsoft.AspNetCore.Http.Connections.Tests/HttpConnectionDispatcherTests.cs:
```diff
@@ -1744,12 +1839,18 @@ private static DefaultHttpContext MakeRequest(string path, ConnectionContext con
                 case HttpTransportType.WebSockets:
-                    context.Features.Set<IHttpWebSocketFeature>(new TestWebSocketConnectionFeature());
+                    var feature = new TestWebSocketConnectionFeature();
+                    if (terminateTransport.CanBeCanceled)
```
Eerr, why?

anurse replied:
I was probably going overboard trying to shut down the mock connection. I can yank it.

davidfowl commented:
> I'm curious why not use CloseMessage?

Decided that this should be at the transport layer since it's impossible to terminate the connection at that level.

moozzyk commented:
`abort` comes back... Can you update the spec?

davidfowl commented on src/Microsoft.AspNetCore.Http.Connections/HttpConnectionDispatcher.cs:
```diff
@@ -487,6 +492,36 @@ private static bool ServerHasWebSockets(IFeatureCollection features)
+            connection.Application.Output.Complete();
+
+            // Dispose the connection gracefully, but don't wait for it
+            _ = connection.DisposeAsync(closeGracefully: true);
```
I think this needs to be call `_ =_manager.DisposeAndRemoveAsync(connection, closeGracefully: false);`, otherwise nobody will remove the disposed connection from the list.

davidfowl commented on test/Microsoft.AspNetCore.Http.Connections.Tests/HttpConnectionDispatcherTests.cs:
```diff
@@ -1616,6 +1616,93 @@ public TrackingMemoryStream(SemaphoreSlim callerTracker, Task waitTask)
+
+                // Verify the response from the DELETE request
+                Assert.Equal(StatusCodes.Status202Accepted, deleteContext.Response.StatusCode);
+                Assert.Equal("text/plain", deleteContext.Response.ContentType);
```
Assert that the connection manager no longer tracks the connection.

davidfowl commented on test/Microsoft.AspNetCore.Http.Connections.Tests/HttpConnectionDispatcherTests.cs:
```diff
@@ -1616,6 +1616,93 @@ public TrackingMemoryStream(SemaphoreSlim callerTracker, Task waitTask)
+                var app = builder.Build();
+                var options = new HttpConnectionOptions();
+
+                var pollTask = dispatcher.ExecuteAsync(context, options, app).OrTimeout();
```
Do you want the timeout here instead of on the awaiting of the pollTask itself?

anurse replied:
We usually do it this way, but I do think it fits better on the `await`. I can put it there.

davidfowl replied:
Yah

BrennanConroy commented on src/Microsoft.AspNetCore.Http.Connections/HttpConnectionDispatcher.Log.cs:
```diff
@@ -40,6 +40,12 @@ private static class Log
                 LoggerMessage.Define(LogLevel.Debug, new EventId(10, "NegotiationRequest"), "Sending negotiation response.");
 
+            private static readonly Action<ILogger, HttpTransportType, Exception> _receivedDeleteRequestForUnsupportedTransport =
+                LoggerMessage.Define<HttpTransportType>(LogLevel.Trace, new EventId(11, "ReceivedDeleteRequestForUnsupportedTransport"), "Recieved DELETE request for unsupported transport: {TransportType}");
```
i before e except after **whenever you feel like it**

anurse replied:
Ugh, that one is literally the bane of my existence.

BrennanConroy replied:
Still need to fix this, also, full stop

BrennanConroy commented on src/Microsoft.AspNetCore.Http.Connections/Internal/Transports/LongPollingTransport.cs:
```diff
@@ -1,13 +1,13 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Logging;
```
sort

anurse commented:
:up: :date: C# client and functional tests, and TypeScript client (no functional tests). Latest PR feedback not yet addressed.

anurse commented:
Something we need to talk about: Should this wait to allow the client to drain messages or is the client expected to never come back after the `DELETE`?

anurse commented:
:up: :date: added support for TypeScript and functional tests. There was a bug in the way transport shutdown occurred on the TypeScript client where the connection didn't await `ITransport.stop` during `HttpConnection.stop`. Also, `HttpConnection.stop` called `HttpConnection.onclose` itself, meaning it could be double-called if `ITransport.onclose` fired. I changed that logic so that `HttpConnection.stop` just stops the transport and returns, and `HttpConnection.onclose` is fired when `ITransport.onclose` fires.

As part of that, I split the transports into three files. I did that in a single commit so if you want to see the TypeScript client changes without the wall of red and green text, look at this commit: [link to come]

davidfowl commented on clients/ts/signalr/src/HttpConnection.ts:
```diff
@@ -5,9 +5,12 @@ import { ConnectionClosed, DataReceived } from "./Common";
 import { DefaultHttpClient, HttpClient } from "./HttpClient";
 import { IConnection } from "./IConnection";
 import { ILogger, LogLevel } from "./ILogger";
+import { ITransport, TransferFormat, TransportType } from "./ITransport";
```
super nice

davidfowl commented on clients/ts/signalr/src/ITransport.ts:
```diff
@@ -0,0 +1,24 @@
+}
+
+export interface ITransport {
+    connect(url: string, transferFormat: TransferFormat, connection: IConnection): Promise<void>;
```
I removed the last argument

davidfowl commented on clients/ts/signalr/src/LongPollingTransport.ts:
```diff
@@ -0,0 +1,166 @@
+            pollOptions.responseType = "arraybuffer";
+        }
+
+        const token = this.accessTokenFactory();
```
where's the await? Did you make these changes before all of the other changes?

anurse replied:
probably, I'll reconcile it

davidfowl replied:
Yes you did you basically don't have changes from the TS changes that were made. The accessToken factory and my change

BrennanConroy commented on clients/ts/signalr/src/ServerSentEventsTransport.ts:
```diff
@@ -0,0 +1,109 @@
+
+                eventSource.onerror = (e: any) => {
+                    const error = new Error(e.message || "Error occurred");
+                    reject(error);
```
Does order matter here? Reject at the end makes more sense to me

anurse replied:
Actually, we should do one or the other. If `onopen` has fired, we resolve the promise (to say we're connected), and we shouldn't then reject it later. If `onopen` hasn't fired, this is an error during the initial connection and we should reject the promise. I'll tidy this up as well.

BrennanConroy commented on clients/ts/signalr/src/Utils.ts:
```diff
@@ -15,3 +18,52 @@ export class Arg {
+    const token = accessTokenFactory();
+    if (token) {
+        headers = {
+            ["Authorization"]: `Bearer ${accessTokenFactory()}`,
```
Use `token`

anurse replied:
derp

anurse commented:
Rebasing on dev to get the missing changes. Not looking forward to that merge :)

anurse commented on clients/ts/signalr/src/HttpConnection.ts:
```diff
@@ -238,22 +242,25 @@ export class HttpConnection implements IConnection {
 
-    private stopConnection(raiseClosed: boolean, error?: Error) {
+        // The transport's onclose will trigger stopConnection which will run our onclose event.
         if (this.transport) {
```
If the transport is null, we haven't started yet. In fact, I should clear out the transport in `stopConnection` so a second call to stop no-ops.

BrennanConroy commented on clients/ts/signalr/src/Utils.ts:
```diff
@@ -15,3 +18,52 @@ export class Arg {
+    if (data instanceof ArrayBuffer) {
+        length = `Binary data of length ${data.byteLength}.`;
+        if (includeContent) {
+            length += ` Content: '${formatArrayBuffer(data)}'.`;
```
Could we remove the '.' from here, logs are now showing up with '..' at the end because the actual log location also puts a '.'

BrennanConroy commented on src/Microsoft.AspNetCore.Http.Connections/HttpConnectionDispatcher.Log.cs:
```diff
@@ -40,6 +40,12 @@ private static class Log
+                LoggerMessage.Define<HttpTransportType>(LogLevel.Trace, new EventId(11, "ReceivedDeleteRequestForUnsupportedTransport"), "Recieved DELETE request for unsupported transport: {TransportType}");
+
+            private static readonly Action<ILogger, Exception> _terminatingConnection =
+                LoggerMessage.Define(LogLevel.Trace, new EventId(12, "TerminatingConection"), "Terminating Long Polling connection due to a DELETE request");
```
full-stop

BrennanConroy commented on test/Microsoft.AspNetCore.Http.Connections.Tests/HttpConnectionDispatcherTests.cs:
```diff
@@ -1616,6 +1616,141 @@ public TrackingMemoryStream(SemaphoreSlim callerTracker, Task waitTask)
+                deleteContext.Request.Path = "/foo";
+                deleteContext.Request.QueryString = new QueryString($"?id={connection.ConnectionId}");
+                deleteContext.Request.Method = "DELETE";
+
```
Assert.False(pollTask.IsCompleted)

JamesNK commented on clients/ts/FunctionalTests/ts/Common.ts:
```diff
@@ -1,31 +1,31 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
-import { IHubProtocol, JsonHubProtocol, TransportType } from "@aspnet/signalr";
+import { IHubProtocol, JsonHubProtocol, HttpTransportType } from "@aspnet/signalr";
```
👍

JamesNK commented on clients/ts/FunctionalTests/ts/HubConnectionTests.ts:
```diff
@@ -559,7 +559,34 @@ describe("hubConnection", () => {
+
+                try {
+                    const hubConnection = new HubConnection("/authorizedhub", {
+                        accessTokenFactory: () => getJwtToken("http://" + document.location.host + "/generateJwtToken"),
```
Is this from the JS async token PR? Curious why it is show up here. Did you move it?

anurse replied:
Because I split `Transports.ts` into four files (one for each transport and one for the interface) it got wiped out in a bad rebase.

JamesNK commented on clients/ts/signalr/src/HttpConnection.ts:
```diff
@@ -5,13 +5,16 @@ import { ConnectionClosed, DataReceived } from "./Common";
 export interface IHttpConnectionOptions {
     httpClient?: HttpClient;
-    transport?: TransportType | ITransport;
+    transport?: HttpTransportType | ITransport;
```
Does TS support flags? Is this a single transport type or multiple? I'm just wondering if this should be puralized like in C# client?

anurse replied:
This wasn't configured as a flags for some reason. I'm not sure if @moozzyk tried it and it didn't work or just didn't do it. I'll file an issue and work it out today.

anurse replied:
https://github.com/aspnet/SignalR/issues/1961

moozzyk replied:
There was no fallback originally so flags did not make much sense.

JamesNK commented on clients/ts/signalr/src/LongPollingTransport.ts:
```diff
@@ -0,0 +1,163 @@
+
+    constructor(httpClient: HttpClient, accessTokenFactory: () => string | Promise<string>, logger: ILogger, logMessageContent: boolean) {
+        this.httpClient = httpClient;
+        this.accessTokenFactory = accessTokenFactory || (() => null);
```
nit: Defaulting to a null string would be a bit more efficient.

anurse replied:
You mean default to `null` rather than a `null`-returning function? We'd have to change the signature and call site.

JamesNK replied:
Ohhhh, I thought it was a string or a func, but it is always a func. nvm

davidfowl commented on src/Microsoft.AspNetCore.Http.Connections.Client/Internal/LongPollingTransport.cs:
```diff
@@ -123,6 +120,9 @@ public Task StartAsync(Uri url, TransferFormat transferFormat)
                 throw;
             }
 
+            _transport.Output.Complete();
```
Why does this matter?

davidfowl replied:
Completing the pipes was causing the poll loop to throw, how?

anurse replied:
> how?

It wasn't throwing, I mistook that. It was simply completing "gracefully" as though the server terminated the poll.

anurse commented:
:up: :date: feedback resolved, holding on Approval and green builds

davidfowl commented on clients/ts/signalr/src/LongPollingTransport.ts:
```diff
@@ -0,0 +1,163 @@
+import { DataReceived, TransportClosed } from "./Common";
+import { HttpError, TimeoutError } from "./Errors";
+import { HttpClient, HttpRequest } from "./HttpClient";
+import { IConnection } from "./IConnection";
```
You can remove these now

davidfowl commented on clients/ts/signalr/src/ITransport.ts:
```diff
@@ -0,0 +1,24 @@
+import { DataReceived, TransportClosed } from "./Common";
+import { IConnection } from "./IConnection";
```
nit: Remove this is unused

davidfowl commented on clients/ts/signalr/src/LongPollingTransport.ts:
```diff
@@ -0,0 +1,174 @@
+                    this.logger.log(LogLevel.Trace, `(LongPolling transport) polling: ${pollUrl}`);
+                    const response = await this.httpClient.get(pollUrl, pollOptions);
+
+                    if (!this.running) {
```
Why do we need this

anurse replied:
As I explained elsewhere, this is just diagnostic. I can yank it.
